File: Python/firstUniqChar2.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: firstUniqChar2.py
@Function: 第一个只出现一次的字符
@Link: https://leetcode-cn.com/problems/di-yi-ge-zhi-chu-xian-yi-ci-de-zi-fu-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-06
"""


from typing import List
import collections


class Solution:
    def firstUniqChar(self, s: str) -> str:
        position = dict()
        q = collections.deque()
        n = len(s)
        for i, ch in enumerate(s):
            if ch not in position:
                position[ch] = i
                q.append((s[i], i))
            else:
                position[ch] = -1
                while q and position[q[0][0]] == -1:
                    q.popleft()
        return ' ' if not q else q[0][0]


# --------------------------
if __name__ == "__main__":
    str = "abaccdeff"

    solution = Solution()
    first_uniq_char = solution.firstUniqChar(str)
    print(f"The solution of this proble is {first_uniq_char}")

File: C++/firstUniqChar2.cpp
/**
 * @File    : firstUniqChar2.cpp
 * @Brief   : 第一个只出现一次的字符
 * @Link    : https://leetcode-cn.com/problems/di-yi-ge-zhi-chu-xian-yi-ci-de-zi-fu-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-06
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <queue>
#include <utility>

class Solution
{
public:
    char firstUniqChar(std::string s)
    {
        std::unordered_map<char, int> position;
        std::queue<std::pair<char, int>> q;
        int n = s.size();
        for (int i = 0; i < n; ++i)
        {
            if (!position.count(s[i]))
            {
                position[s[i]] = i;
                q.emplace(s[i], i);
            }
            else
            {
                position[s[i]] = -1;
                while (!q.empty() && position[q.front().first] == -1)
                {
                    q.pop();
                }
            }
        }
        return q.empty() ? ' ' : q.front().first;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    std::string str = "abaccdeff";

    Solution solution;
    auto first_uniq_char = solution.firstUniqChar(str);
    std::cout << "The solution of this problem is " << first_uniq_char << std::endl;

    return 0;
}
